design/service-apis.md

```diff
@@ -0,0 +1,389 @@
+with `controller: projectcontour.io/contour-operator` will be managed by Contour Operator. There MUST be at least one
+GatewayClass defined in order to have a functional Gateway. This is similar to IngressClass for Ingress and StorageClass
+for PersistentVolumes.
+
+The following is an example of a GatewayClass managed by Contour Operator:
+
+```yaml
+kind: GatewayClass
+apiVersion: networking.x-k8s.io/v1alpha1
+metadata:
+  name: example-gatewayclass
+spec:
+  controller: projectcontour.io/contour-operator
+  parametersRef:
+    name: example-contour
+    group: operator.projectcontour.io
+    kind: Contour
+```
+
+The `controller` field is a domain/path string that indicates the controller responsible for managing Gateways of this
+class. The `projectcontour.io/contour-operator` string will be used to specify Contour Operator as the responsible
+controller. This field is not mutable and cannot be empty.
+
+The `parametersRef` field is a reference to an implementation-specific resource containing configuration parameters
+associated to the GatewayClass. This field will be used to expose Contour-specific configuration. If the field is
+unspecified, the operator will create an instance of Contour with default settings. Since `parametersRef` does not
+contain a namespace field, the referenced resource must exist in a known namespace. The operator's namespace will be
+used as the known namespace. If the referenced resource cannot be found, the operator will set the GatewayClass
+`InvalidParameters` status condition to true.
+
+### Contour
+
+The Contour custom resource will be used to provide a structured object for representing the configuration of a Contour
+instance. A Contour custom resource will be consumed by a GatewayClass. Since the Contour custom resource currently
+triggers the operator to instantiate a Contour instance, a field will be added to prevent this behavior and treat the
+Contour as a configuration-only resource. The following data structure will be used to manage this behavior:
+
+```go
+// ContourSpec currently exists.
+type ContourSpec struct {
+	// GatewayClassRef is a reference to a GatewayClass name used for
+	// managing a Contour.
+	//
+	// +kubebuilder:default=None
+    GatewayClassRef string `json:"gatewayClassRef,omitempty"`
+}
+```
+
+The `gatewayClassRef` field is a reference to a GatewayClass used for managing the Contour. If unspecified, the field
+will default to "None", indicating a GatewayClass does not exist, and the operator should use the Contour resource to
+instantiate a Contour instance.
+
+The following is an example of a Contour resource that references a GatewayClass:
+```yaml
+apiVersion: operator.projectcontour.io/v1alpha1
+kind: Contour
+metadata:
+  name: example-contour
+  namespace: contour-operator
+spec:
+  gatewayClassRef: example-gatewayclass
+  replicas: 2
+  namespace:
+    name: projectcontour
```

Comment:
Since a `Contour` defines a namespace, does this mean that each `GatewayClass` can only be instantiated in a single namespace?

What happens if a user defines a `Gateway` for this `GatewayClass`, in a namespace other than `projectcontour`?

More generally, I wonder if there are, or will be, fields in the `Contour` CRD spec that won't make sense as pre-defined parameters for a `GatewayClass`, and if so, whether we'd want to use something other than a `Contour` CR to define `GatewayClass` config.

Reply:
> Since a Contour defines a namespace, does this mean that each GatewayClass can only be instantiated in a single namespace?

@skriss GatewayClass is a cluster-scoped resource.

> What happens if a user defines a Gateway for this GatewayClass, in a namespace other than projectcontour?

@skriss currently, `gatewayclass.spec.parametersRef` should reference a cluster-scoped resource. This will be changing in the future to support cluster-scoped and namespaced-scoped resource references. We'll see if the parametersRef needs to be in the same namespace as Gateways of this GatewayClass. In the meantime, I suggest we don't impose such restriction. Conceivably, one persona can be responsible for managing cluster-wide proxy configs, i.e. GatewayClasses/Contours and a different persona be responsible for managing Gateways.

> More generally, I wonder if there are, or will be, fields in the Contour CRD spec that won't make sense as pre-defined parameters for a GatewayClass, and if so, whether we'd want to use something other than a Contour CR to define GatewayClass config.

I originally went in the direction by proposing the `ContourConfig` config resource for Service APIs. @jpeach suggested that we continue using the @Contour resource and I tend to agree with him. Until we have a solid use-case for not using the Contour CRD, I suggest we stick with it.

Reply:
With respect to having a Gateway-specific thing, I think it's fine to leave the `Contour` object reference to start with, we can change our minds to another thing later. I think that any other thing would be substantially similar to the `Contour`, so we'd probably end up needing to do an embedded struct or something anyway, to cover the similarities.